### ISSUE TITLE
FFmpegFrameRecorder trim audio file when converting to PCM_ALAW

### ISSUE THREAD

**Issue by canelzio** [bytedeco/javacv#876]
Hello,
I'm trying to implement an audio converter using FFmpegFrameGrabber and FFmpegFrameRecorder.
Everything seems to work fine, but when I try to convert an AMR-NB file to PCM_ALAw I see that resulted audio file has a smaller duration than the original one (it seems to be trimmed of about 500ms).
Convertion to other codecs does not seem to be affected by the problem.

Following the extract of the simple code I'm using:
```
			FFmpegFrameGrabber grabber = new FFmpegFrameGrabber(inFile);
			grabber.start();

			FFmpegFrameRecorder recorder = new FFmpegFrameRecorder(outFile, grabber.getAudioChannels());
			recorder.setSampleRate(8000);
			recorder.setFormat("wav");
			recorder.setAudioCodec(org.bytedeco.javacpp.avcodec.AV_CODEC_ID_PCM_ALAW);
			
		        recorder.start();
			
			Frame frame = null;
			while ((frame = grabber.grabFrame()) != null) {
				recorder.record(frame);
			}
```

I also tried to execute the same convertion (AMR-NB to PCM_ALAW) using ffmpeg command line utility and it works perfectly:

`ffmpeg -v debug -i GetGreeting.amr -c pcm_alaw -f wav -ar 8000 GetGreeting.alaw`

Debug messages from ffmpeg shows that it uses by default some filters that JavaCV does not (don't know if it could impact):

_[format_out_0_0 @ 0x2f25a80] auto-inserting filter 'auto_resampler_0' between the filter 'Parsed_anull_0' and the filter 'format_out_0_0'
[AVFilterGraph @ 0x2f250a0] query_formats: 4 queried, 6 merged, 3 already done, 0 delayed
[auto_resampler_0 @ 0x2f27fe0] [SWR @ 0x2f28400] Using fltp internally between filters
[auto_resampler_0 @ 0x2f27fe0] ch:1 chl:mono fmt:flt r:8000Hz -> ch:1 chl:mono fmt:s16 r:8000Hz_

Analyzing debug messages from ffmpeg I also noted that it processed 21440 samples, while resulted file from javacv has only 16384.

I would like to replicate exactly the same behavior of ffmpeg via javacv, but maybe I'm missing some configuration.
Hope you could help.

Thank you.

**Comment by saudet** [comment]
FFmpegFrameRecorder uses swresample to do the resampling, not avfilter. Maybe there is a bug in swresample. You could try to use avfilter as with the ffmpeg program by using FFmpegFrameFilter.

**Comment by canelzio** [comment]
Thanks Samuel for suggestion, but I'm not able to use FFmpegFrameFilter for audio streams.
This line:
`FFmpegFrameFilter filter = new FFmpegFrameFilter("aresample", 0, 0);`

produce the error:
_[buffer @ 0x7ff9fc13c800] Unable to parse option value "0x0" as image size
[buffer @ 0x7ff9fc13c800] Error setting option video_size to value 0x0.
[in @ 0x7ff9fc13c6e0] Error applying options to the filter._

And this:
`FFmpegFrameFilter filter = new FFmpegFrameFilter("aresample", 1, 1);`

the following:
_[in @ 0x7ffa280e7a60] Media type mismatch between the 'in' filter output pad 0 (video) and the 'Parsed_aresample_0' filter input pad 0 (audio)
Cannot create the link buffer:0 -> aresample:0_

Could you please address me with the correct FFmpegFrameFilter usage for audio streams only?
Thank you in advance.

**Comment by saudet** [comment]
Ah, you're right. FFmpegFrameFilter doesn't support audio frames yet. I'll mark this as a duplicate of issue #492. Would you like to try implementing the support? I'll help you with any issues you encounter.

**Comment by canelzio** [comment]
Hi Samuel,
I tried to implement such new support, but for me (I’m not an expert of filtering, re-sampling...) it is not so easy. If you want I can send you the work I’ve done so far, hoping it could be useful.

Anyway I think I found a bug on the FFmpegFrameRecorder.record() method: it seems to ignore final samples in the last frame.
So I think to add a record(AVFrame) call in the stop() method if needed.

In this way all samples are written in the output file and it is no more truncated. This solved for me.

Let me now if you are interested in my work and how you prefer to manage it.

Thank you.
Regards.
